go/fn/object.go

```diff
@@ -759,6 +784,107 @@ func (o KubeObjects) Less(i, j int) bool {
 	return idStrI < idStrJ
 }
 
+func (o KubeObjects) String() string {
+	var elems []string
+	for _, obj := range o {
+		elems = append(elems, strings.TrimSpace(obj.String()))
+	}
+	return strings.Join(elems, "\n---\n")
+}
+
+// Select will return the subset of objects in KubeObjects such that f(object) returns 'true'.
+func (o KubeObjects) Select(f func(o *KubeObject) bool) KubeObjects {
+	var result KubeObjects
+	for _, obj := range o {
+		if f(obj) {
+			result = append(result, obj)
+		}
+	}
+	return result
+}
+
+// SelectByGvk will return the subset of objects that matches the provided GVK.
+func (o KubeObjects) SelectByGvk(apiVersion, kind string) KubeObjects {
+	return o.Select(func(o *KubeObject) bool {
+		return o.IsGVK(apiVersion, kind)
+	})
+}
+
+// ExcludeByGvk will return the subset of objects that do not match the provided GVK.
+func (o KubeObjects) ExcludeByGvk(apiVersion, kind string) KubeObjects {
+	return o.Select(func(o *KubeObject) bool {
+		return !o.IsGVK(apiVersion, kind)
+	})
+}
+
+// SelectByName will return the subset of objects that matches the provided name.
+func (o KubeObjects) SelectByName(name string) KubeObjects {
+	return o.Select(func(o *KubeObject) bool {
+		return o.GetName() == name
+	})
+}
+
+// ExcludeByName will return the subset of objects that do not match the provided name.
+func (o KubeObjects) ExcludeByName(name string) KubeObjects {
+	return o.Select(func(o *KubeObject) bool {
+		return o.GetName() != name
+	})
+}
+
+// SelectByNamespace will return the subset of objects that matches the provided namespace.
+func (o KubeObjects) SelectByNamespace(namespace string) KubeObjects {
+	return o.Select(func(o *KubeObject) bool {
+		return o.GetNamespace() == namespace
+	})
+}
+
+// ExcludeByNamespace will return the subset of objects that do not match the provided namespace.
+func (o KubeObjects) ExcludeByNamespace(namespace string) KubeObjects {
+	return o.Select(func(o *KubeObject) bool {
+		return o.GetNamespace() != namespace
+	})
+}
+
+// SelectByLabels will return the subset of objects that matches the provided labels.
+func (o KubeObjects) SelectByLabels(labels map[string]string) KubeObjects {
+	return o.Select(func(o *KubeObject) bool {
+		return o.HasLabels(labels)
+	})
+}
+
+// ExcludeByLabels will return the subset of objects that do not match the provided labels.
+func (o KubeObjects) ExcludeByLabels(labels map[string]string) KubeObjects {
```

Comment:
The semantics here are a little tricky - is this excluding where any labels match or all labels match?  (I do know the answer :-) ).  This is where _not_ defining a helper method might well be easier because then we can let users call `Where(Not(HasLabels(...)))` and `Where(Not(HasAnyLabel(...)))`.

Reply:
That makes sense, thanks!